doc/design/v2_layer.md

```diff
@@ -0,0 +1,126 @@
+# Using tape to refactor current paddle.v2 configuration parsing
+
+## What are the problems
+
+Paddle V2 API give a flexible way to configure neural network topology. The user can create a neural network topology layer by layer. We use the final layer to represent the neural network topology.
+
+The example code for users is:
+
+```python
+img = data_layer(name="img", type=Dense(784))
+hidden = fc_layer(input=img, size=200)
+prediction = fc_layer(input=hidden, size=10, act=Softmax())
+cost = classification_cost(input=prediction, label=data_layer(name="label", type=Integer(10)))
+
+paddle.train(cost, ...)
+```
+
+We use `cost` to represent the entire topology of the neural network.  We use the last node to traverse the entire topology by `depth first search` algorithm. The connection between each layer is represented by the `input` parameter.  It is fit for representing a plain neural network topology. However, there are some special layers in Paddle, which are not connected explicitly by `input` parameter. They are:
+
+* Evaluator.  An evaluator is used to compute metrics(such as error rate, f1 score) in Paddle. An evaluator can not be the input of other layers. So we cannot access evaluators by simply traversing back from the final layer.
```

Comment:
There is a broken logic here. From the first statement

> An evaluator can not be the input of other layers. 

we cannot derive the conclusion

> So we cannot access evaluators by simply traversing back from the final layer.

because logically an evaluator could be the "final layer".